cmake/modules/yaml.cmake

```diff
@@ -0,0 +1,403 @@
+  zephyr_check_arguments_required(${CMAKE_CURRENT_FUNCTION} ARG_YAML NAME)
+  yaml_context(EXISTS NAME ${ARG_YAML_NAME} result)
+
+  if(result)
+    message(FATAL_ERROR "YAML context '${ARG_YAML_NAME}' already exists."
+            "Please create a YAML context with a unique name"
+    )
+  endif()
+endfunction()
+
+# Usage
+#   yaml_context(EXISTS NAME <name> <result>)
+#
+# Function to query the status of the YAML context with the name <name>.
+# The result of the query is stored in <result>
+#
+# EXISTS     : Check if the YAML context exists in the current scope
+#              If the context exists, the True is returned in <result>
+# NAME <name>: Name of the YAML context
+# <result>   : Variable to store the result of the query.
+#
+function(yaml_context)
+  cmake_parse_arguments(ARG_YAML "EXISTS" "NAME" "" ${ARGN})
+  zephyr_check_arguments_required_all(${CMAKE_CURRENT_FUNCTION} ARG_YAML EXISTS NAME)
+
+  if(NOT DEFINED ARG_YAML_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "Missing argument in "
+            "${CMAKE_CURRENT_FUNCTION}(EXISTS NAME ${ARG_YAML_NAME} <result-var>)."
+    )
+  endif()
+
+  if(TARGET ${ARG_YAML_NAME}_scope)
+    list(POP_FRONT ARG_YAML_UNPARSED_ARGUMENTS out-var)
+    set(${out-var} TRUE PARENT_SCOPE)
+  else()
+    set(${out-var} ${ARG_YAML_NAME}-NOTFOUND PARENT_SCOPE)
+  endif()
+endfunction()
+
+# Usage:
+#   yaml_create(FILE <file> NAME <name>)
+#
+# Create a new empty YAML contextx.
+# Use the file <file> for storing the context when 'yaml_write(NAME <name>)i is
+# called.
+#
+# Values can be set by calling 'yaml_set(NAME <name>)') by using the <name>
+# specified when creating the YAML context.
+#
+# FILE <file>: Path to file to be used together with this YAML context.
+# NAME <name>: Name of the YAML context.
+#
+function(yaml_create)
+  cmake_parse_arguments(ARG_YAML "" "FILE;NAME" "" ${ARGN})
+
+  zephyr_check_arguments_required_all(${CMAKE_CURRENT_FUNCTION} ARG_YAML FILE NAME)
+
+  internal_yaml_context_free(NAME ${ARG_YAML_NAME})
+  zephyr_create_scope(${ARG_YAML_NAME})
+  zephyr_set(FILE ${ARG_YAML_FILE} SCOPE ${ARG_YAML_NAME})
+  zephyr_set(JSON "{}" SCOPE ${ARG_YAML_NAME})
+endfunction()
+
+# Usage:
+#   yaml_load(FILE <file> NAME <name>)
+#
+# Load an existing YAML file and store its content in the YAML context <name>.
+#
+# Values can later be retrieved (yaml_get()) or set/updated (yaml_set()) be using
+# the same YAML scope name.
+#
+# FILE <file>: Path to file to load.
+# NAME <name>: Name of the YAML context.
+#
+function(yaml_load)
+  cmake_parse_arguments(ARG_YAML "" "FILE;NAME" "" ${ARGN})
+
+  zephyr_check_arguments_required_all(${CMAKE_CURRENT_FUNCTION} ARG_YAML FILE NAME)
+  internal_yaml_context_free(NAME ${ARG_YAML_NAME})
+
+  zephyr_create_scope(${ARG_YAML_NAME})
+  zephyr_set(FILE ${ARG_YAML_FILE} SCOPE ${ARG_YAML_NAME})
+
+  execute_process(COMMAND python -c
+    "import json; import yaml; print(json.dumps(yaml.safe_load(open('${ARG_YAML_FILE}'))))"
```

Comment:
(not requesting change) something to probably thing about is that this escapes from of cmake into python as another process, from a security perspective anyone can put any file that is visible to the OS and might allow some level of privileged file access

Reply:
The `ARG_YAML_FILE` is passed as an argument to `yaml_load()`, so true that a caller of said function can do:
`yaml_load(NAME foo FILE "<insert python escape command>")`.

But if the caller of `yaml_load()` can do that, then that caller can much easier call `execute_process()` directly without the hassle of going through another function. Just call: `execute_process(COMMAND python -c "...")` directly.

So I guess you're basically saying that CMake is unsafe, especially if you use `sudo cmake` :wink: